### ISSUE TITLE
Login authentication can be bypassed by loading resources dynamically

### ISSUE THREAD

**Issue by WiXSL** [marmelab/react-admin#5374]
I followed the docs [Declaring resources at runtime](https://marmelab.com/react-admin/Admin.html#declaring-resources-at-runtime) and found this:

1.- Logout
2.- Inside the sandbox address bar, remove by hand the "login" out of the URL.

Result: You access the app without credentials.

https://codesandbox.io/s/solitary-snow-zvwnr

**Comment by djhi** [comment]
This is weird, we actually use runtime declaration of resources in the simple example, and we are redirected to the login page if not authenticated

**Comment by WiXSL** [comment]
This is the simple example, I only modify index.js changing the resources for the `fetchResources` function.

**Comment by fzaninotto** [comment]
That's because you only have empty resources, so react-admin creates no content and therefore makes no auth check.

The auth check is done on each CRUD route (e.g. /posts, /posts/:id, /posts/create, etc). As your resources are empty, there is no CRUD route. If you add a `list={ListGuesser}` to your resources, you will see that the user is properly redirected to the login when trying to access inner pages.  

So basically your codeSandbox shows that a non-identified user can access an empty app in react-admin. I don't see the problem with that.

Or am I missing something?

**Comment by WiXSL** [comment]
@fzaninotto,
I've added `list={ListGuesser}` to the dynamic resources and the bug is still reproducible

https://codesandbox.io/s/nice-archimedes-prg5w

**Comment by fzaninotto** [comment]
Something is wrong with your last sandbox. The exact same code (I forked yours) doesn't show the problem anymore: https://codesandbox.io/s/keen-nobel-3njmc

**Comment by WiXSL** [comment]
You are right!
The use case I was trying to show in the case an error occurs during fetching in this scenario like in this code sandbox: https://codesandbox.io/s/nostalgic-darkness-tmg5f

You don't get react-admin cleaning, redirection and notification.
Maybe the docs should be enhanced by showing how to check authorization and redirect manually to the login for this use case.

**Comment by fzaninotto** [comment]
Are you referring to [the "Catching Authentication Errors On The API" section](https://marmelab.com/react-admin/Authentication.html#catching-authentication-errors-on-the-api) in the authentication documentation? this seems pretty clear to me.

**Comment by WiXSL** [comment]
Of course, my point is that  in the case of loading resources dynamically (https://codesandbox.io/s/nostalgic-darkness-tmg5f) react-admin doesn't  call the `authProvider.checkError()` method when a 401 error is thrown

**Comment by fzaninotto** [comment]
Well, react-admin has no way of knowing you're fetching data in the admin child, and besides you're doing it using `fetch`, not the `dataProvider`. Therefore it's your responsibility to check for errors there.

**Comment by WiXSL** [comment]
Sure, currently I'm checking for errors manually in my code. 
I'm just trying to load resources from the server dynamically plus having react-admin authentication functionality as well.

I'll try to summarize:

- [Declaring resources at runtime](https://marmelab.com/react-admin/Admin.html#declaring-resources-at-runtime)
  Has the drawback of not having react-admin's self-aware authorization functionality plus not being able to use hooks. 
 
- [Unplugging the Admin using AdminContext and AdminUI](https://marmelab.com/react-admin/Admin.html#unplugging-the-admin-using-admincontext-and-adminui) 
  Has the problem of  loading resources once the login page is render and not on Admin's render (which is incorrect as an example of loading resources dynamically, right?)

Right now, I've solved it by using [Declaring resources at runtime](https://marmelab.com/react-admin/Admin.html#declaring-resources-at-runtime) plus manual authentication like you suggested.
But it would be great to have a way react-admin to do this with the best of both options.
Maybe the new <Admin/> ready prop could help. I'll have to try it out.

**Comment by fzaninotto** [comment]
In your case, you clearly have to use the "Unplugging the Admin using AdminContext and AdminUI" path.

However, I can't reproduce the problem you describe. I don't see the login page during loading. But I see the 404 page during loading, which is another problem but that is easy to fix.

Could you post a CodeSandbox demonstrating the issue? Here is the code I'm using:

```jsx
/* eslint react/jsx-key: off */
import * as React from 'react';
import { useState, useEffect } from 'react';
import { AdminContext, AdminUI, Resource } from 'react-admin'; // eslint-disable-line import/no-unresolved
import { render } from 'react-dom';

import authProvider from './authProvider';
import comments from './comments';
import dataProvider from './dataProvider';
import i18nProvider from './i18nProvider';
import Layout from './Layout';
import posts from './posts';
import users from './users';
import tags from './tags';

const resources = [
    <Resource name="posts" {...posts} key="posts" />,
    <Resource name="comments" {...comments} key="comments" />,
    <Resource name="users" {...users} key="users" />,
    <Resource name="tags" {...tags} key="tags" />,
];

const AsyncResources = () => {
    const [resourcesLoaded, setResourcesLoaded] = useState(false);
    useEffect(() => {
        setTimeout(() => setResourcesLoaded(true), 1500);
    });
    const resourcesToRender = resourcesLoaded ? resources : [];
    return <AdminUI>{resourcesToRender.map(resource => resource)}</AdminUI>;
};

render(
    <AdminContext
        authProvider={authProvider}
        dataProvider={dataProvider}
        i18nProvider={i18nProvider}
        title="Example Admin"
        layout={Layout}
    >
        <AsyncResources />
    </AdminContext>,
    document.getElementById('root')
);
```

**Comment by WiXSL** [comment]
Using your code, ando starting at Login page, the menus get loaded there.
I need themm to load once the user is loged in.

https://codesandbox.io/s/loving-meninsky-sdle0?file=/src/index.js

**Comment by fzaninotto** [comment]
@WiXSL This is not something supported out of the box by react-admin, you'll have to implement your own menu for that.
